contracts/CATFactory.sol

```diff
@@ -44,19 +45,29 @@ contract CATFactory is Ownable {
         address catAddress = address(newCAT);
         administerableTokens[msg.sender].push(catAddress);
         emit CATCreated(msg.sender, catAddress, _nextTokenId);
-        _nextTokenId++; // Increment tokenId for the next contract
+        _nextTokenId++;
 
         return catAddress;
     }
 
     /**
-     * @dev Grants minter role to an address in the CAT contract.
-     * @param catAddress The address of the CAT contract.
-     * @param minter The address to grant the minter role.
+     * @dev Notifies the factory that a minter role has been granted in a CAT contract.
+     * This function is called by CAT contracts when their admins grant minter roles.
+     * @param catAddress The address of the CAT contract calling this function.
+     * @param minter The address that was granted the minter role.
      */
-    function grantMinterRole(address catAddress, address minter) public onlyOwner {
-        ContributionAccountingToken(catAddress).grantMinterRole(minter);
-        mintableTokens[minter].push(catAddress); // Update mintable tokens mapping
+    function onMinterRoleGranted(address catAddress, address minter) external {
+        // Verify that the caller is the CAT contract
+        require(msg.sender == catAddress, "Only CAT contract can call this function");
+        
+        // Check if catAddress already exists in the array
+        address[] storage minterTokens = mintableTokens[minter];
+        for (uint256 i = 0; i < minterTokens.length; i++) {
```

Comment:
This loop can become too expensive when `minterTokens` becomes too long. Instead, use a `mapping(address => mapping(address => bool)` to store whether a given address has already been registered as a minter for a given CAT. Then the check can be done in constant time instead of linear time.